index.js

```diff
@@ -58,6 +58,122 @@ class DbFacility extends Base {
+   * @param {(err: Error|undefined) => void} cb
+   */
+  runTransaction (func, cb) {
+    /** @type {mysql.PoolConnection} */
+    let conn = null
+    let txStarted = false
+    let txCommited = false
+
+    async.series({
+      conn: (next) => {
+        this.cli.getConnection((err, res) => {
+          if (err) return next(err)
+
+          conn = res // intentional
+          return next()
+        })
+      },
+      begin: (next) => {
+        conn.beginTransaction(next)
+      },
+      execute: (next) => {
+        txStarted = true
+        func({ cli: conn }, next)
+      },
+      commit: (next) => {
+        conn.commit(next)
+      },
+      release: (next) => {
+        txCommited = true
+        try {
+          conn.release()
+          return next()
+        } catch (err) {
+          return next(err)
+        }
+      }
+    }, (txFailureErr) => {
+      if (!txFailureErr) return cb()
+
+      if (txStarted && !txCommited) {
+        return async.series({
+          rollback: (next) => {
+            conn.rollback(next)
+          },
+          release: (next) => {
+            try {
+              conn.release()
+              return next()
+            } catch (err) {
+              return next(err)
+            }
+          }
+        }, (rollbackErr) => {
+          if (!rollbackErr) {
+            return cb(txFailureErr)
+          }
+
+          try {
+            conn.destroy() // force cleanup session
+            return cb(txFailureErr)
+          } catch (destroyErr) {
+            return cb(destroyErr)
+          }
+        })
+      }
+
+      return cb(txFailureErr)
+    })
+  }
+
+  /**
+   * @see https://www.npmjs.com/package/mysql#pooling-connections
+   * @see https://www.npmjs.com/package/mysql#transactions
+   *
+   * @param {(conn: { queryAsync: Function, cli: mysql.PoolConnection }) => Promise<void>} func
+   */
+  async runTransactionAsync (func) {
+    /** @type {mysql.PoolConnection} */
+    let conn = null
+    let txStarted = false
+    let txCommited = false
+
+    try {
+      conn = await new Promise((resolve, reject) => {
+        this.cli.getConnection((err, cli) => err ? reject(err) : resolve(cli))
+      })
+      const queryAsync = promisify(conn.query.bind(conn))
+
+      await new Promise((resolve, reject) => conn.beginTransaction((err) => err ? reject(err) : resolve()))
+      txStarted = true
+
+      await func({ queryAsync, cli: conn })
+
+      await new Promise((resolve, reject) => conn.commit((err) => err ? reject(err) : resolve()))
+      txCommited = true
+
+      conn.release()
+    } catch (err) {
+      if (txStarted && !txCommited) {
+        try {
+          await new Promise((resolve, reject) => conn.rollback((err) => err ? reject(err) : resolve()))
+          conn.release()
+        } catch (err) {
+          conn.destroy() // force cleanup session
+        }
+      }
+
+      throw err
```

Comment:
methinks we shouldn't throw `err` if `txCommited` is `true` since it indicates `func` ran successfully. Throwing this error forces the transaction runner to care about an error they need not care about.

Reply:
hmm maybe we can implement a custom error class that has info also about transaction state? e.g. something like this:
`throw new MysqlTransactionError(msg, originalError, txStarted, txCommited)`

Reply:
Hmm, this would work but it seems to me that returning an error that occurs after the transaction has committed will make the API too hard to use. Are there any upsides of burdening the user of the API about errors that occur once the transaction is committed rather than handling those errors in the library?

Reply:
Well I prefer to leave full responsibility to caller to handle errors, by giving this custom error and providing info about it in docs seems enough for usage

Reply:
That's fine.